Fix resolve_version_query to not kill tee in filter_version_candidates

M1kep opened the pull request:
Prior to this change the awk 'prog' contained an exit command. When tee attempted to write to the stdout at this point, it received a SIGPIPE signal and exited with an error code of 141. I can't seem to find a way to have tee ignore this signal, but I don't believe the overhead of awk doing a small comparison will have a large/any performance impact especially relative to the processing in filter_version_candidates

Fixes #263

Stratus3D commented:
Hi @M1kep, thanks for the PR. I am not sure I understand what the issue is and how this fixes it. By default awk's `exit` will cause awk to exit with a status of 0 (successful). It's not clear to me how this changes the behavior of awk, other than not exiting when the first match is found.

M1kep commented:
Hi @Stratus3D , happy to try to explain the best I can. Admittedly, I can't reproduce this behavior outside of this projects so I apparently am missing something. But here's what I know:

Logging Changes I made for examples:
Added `    echo "Version: $version">&2` at https://github.com/asdf-vm/asdf-nodejs/blob/master/lib/utils.sh#L43
Added `    echo ${PIPESTATUS[@]}>&2` within the else at https://github.com/asdf-vm/asdf-nodejs/blob/master/lib/utils.sh#L84

Current behavior:
```
❯ asdf install nodejs 16.8.0
Version: 16.10.0
Version: 16.9.1
Version: 16.9.0
Version: 16.8.0
Version: 16.7.0
Version: 16.6.2
141 141 141
... Removed Download Output ...
```

Currently when a specific version is installed, the [filter_version_candidates](https://github.com/asdf-vm/asdf-nodejs/blob/master/lib/utils.sh#L31) function is only processing the version up to the requested version, and two version after that. The exit codes of `141`(`SIGPIPE`) indicate that the three programs in the pipeline are all exiting due to a `SIGPIPE` error. My understanding is that because `awk` is exiting when it finds a match, the next time `tee` tries to write to `STDOUT`, it triggers a `SIGPIPE` error as `awk` has exited and is no longer listening.

Because of this early exit, the cache($HOME/.asdf/tmp/nodejs/cache/index) looks like this:

```
16.10.0	16.10.0
16.9.1	16.9.1
16.9.0	16.9.0
16.8.0	16.8.0
16.7.0	16.7.0
```

This PR fixes this by keeping the `awk` process alive until it exits naturally. With the same two logging lines added in addition to the changes proposed in this PR, the below is what happens:

```
❯ asdf install nodejs 16.8.0
Version: 16.10.0
Version: 16.9.1
Version: 16.9.0
Version: 16.8.0
Version: 16.7.0
Version: 16.6.2
Version: 16.6.1
... removed for ease of viewing ...
Version: 0.1.18
Version: 0.1.17
Version: 0.1.16
Version: 0.1.15
Version: 0.1.14
0 0 0
```

Now, all of the versions are processed, and the pipeline exits with `0` as expected. Additionally, the config file is also properly populated so future calls such as `asdf list all nodejs` return the correct results.

augustobmoura commented:
@M1kep, do you want to open a new PR for that? I think it should be easy fix

M1kep commented:
@augustobmoura I'm happy to open a PR, are you suggesting replacing tee with a > "$index_file" and then `cat`ting that file separately(On the next line)
